### PR TITLE
feat(storage): common interface and structural change for functional refactor

### PR DESCRIPTION
**Pull request by AllanZhengYP**
#### Description of changes

* Add basic types for functional storage API
* Add placeholder for implementation of the APIs.

#### Checklist

- [x] PR description included
- [ ] `yarn test` passes
- [ ] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/storage/src/types/params.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import { FileDownloadOptions, StorageOperationParameter } from './common';
+
+export type StorageOptions = { level?: 'guest' } & {
+	level?: 'protected' | 'private';
+	identityId?: string;
+};
```
Comment (kvramyasri7) [inline packages/storage/src/types/params.ts:9]:
```suggestion
export type StorageOptions = { level?: 'guest'  |  'protected'} | {
	level:  'private';
	identityId: string;
};
```

File: packages/storage/src/providers/s3/types/results.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+	 */
+	contentLength?: number;
+	/**
+	 * An entity tag (ETag) is an opaque identifier assigned by a web server to a specific version of a resource found at a URL.
+	 */
+	eTag?: string;
+	/**
+	 * The tag-set for the object. The tag-set must be encoded as URL Query parameters. (For example, "Key1=Value1")
+	 * @see https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-tagging.html
+	 */
```
Comment (kvramyasri7) [inline packages/storage/src/providers/s3/types/results.ts:26]:
```suggestion
	/**
	 * The user-defined metadata for the object uploaded to S3.
	 * @see https://docs.aws.amazon.com/AmazonS3/latest/userguide/UsingMetadata.html#UserMetadata
	 */
```

File: packages/storage/src/providers/s3/apis/downloadData.ts [hunk 3]
```diff
@@ -0,0 +1,10 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import { StorageOperationParameter, DownloadTask } from '../../../types';
```
Comment (kvramyasri7) [inline packages/storage/src/providers/s3/apis/downloadData.ts:4]:
```suggestion
import { StorageDownloadDataParameter, DownloadTask } from '../../../types';
```

File: packages/storage/src/providers/s3/apis/downloadData.ts [hunk 4]
```diff
@@ -0,0 +1,10 @@
...
+import { S3TransferOptions, S3DownloadDataResult } from '../types';
+
+// TODO: pending implementation
+export declare const downloadData: (
+	params: StorageOperationParameter<S3TransferOptions>
```
Comment (kvramyasri7) [inline packages/storage/src/providers/s3/apis/downloadData.ts:9]:
```suggestion
	params: StorageDownloadDataParameter<S3TransferOptions>
```

File: packages/storage/src/providers/s3/types/options.ts [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+	validateObjectExistence?: boolean;
+	/**
+	 * Whether to use accelerate endpoint.
+	 * @default false
+	 */
+	useAccelerateEndpoint?: boolean;
+};
+
+/**
+ * Parameter options interface for S3 downloadData, downloadFile, uploadData, uploadFile APIs.
```
Comment (kvramyasri7) [inline packages/storage/src/providers/s3/types/options.ts:21]:
```suggestion
 * Parameter options type for S3 downloadData, downloadFile, uploadData, uploadFile APIs.
```

File: packages/storage/src/providers/s3/types/options.ts [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+	 */
+	onProgress?: (event: TransferProgressEvent) => void;
+};
+
+export type S3GetUrlOptions = S3Options & {
+	/**
+	 * Number of seconds till the URL expires.
+	 * @default 900 (15 minutes)
+	 */
+	expiration?: number;
```
Comment (kvramyasri7) [inline packages/storage/src/providers/s3/types/options.ts:35]:
```suggestion
	expiration?: number;
	/**
	 * The default content-disposition header value of the file.
	 * @see https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Disposition
	 */
	contentDisposition?: string;
	/**
	 * The default content-encoding header value of the file.
	 * @see https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Encoding
	 */
	contentEncoding?: string;
		/**
	 * The default content-language header value of the file.
	 * @see https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Language
	 */
	contentLanguage? string;
	/**
	 * The default content-type header value of the file.
	 * @see https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Type
	 */
	contentType?: string;
```

File: packages/storage/src/providers/s3/types/results.ts [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+
+import {
+	StorageDownloadDataResult,
+	StorageGetUrlResult,
+	StorageUploadResult,
+} from '../../../types';
+
+type S3ObjectInformation = {
+	/**
+	 * Creation date of the object.
```
Comment (kvramyasri7) [inline packages/storage/src/providers/s3/types/results.ts:12]:
```suggestion
	 * Last Modified date of the object.
```

Comment (AllanZhengYP) [reply packages/storage/src/providers/s3/types/results.ts:12]: @kvramyasri7 Surprising as it  is, but it's from the [S3 API reference](https://docs.aws.amazon.com/AmazonS3/latest/API/API_GetObject.html#:~:text=Creation%20date%20of%20the%20object.)

File: packages/storage/src/providers/s3/apis/copy.ts [hunk 8]
```diff
@@ -0,0 +1,5 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+// TODO: pending implementation
+export declare const copy: (params: any) => Promise<any>;
```
Comment (Samaritan1011001) [inline packages/storage/src/providers/s3/apis/copy.ts:5]: `any` is a placeholder and will be replaced with specific types later?

File: packages/storage/src/types/common.ts [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+export type TransferProgressEvent = {
+	target?: TransferTask<any>;
+	// Transferred data in bytes
+	transferred: number;
+	// Total data in bytes
+	total?: number;
+};
+
+export type UploadTask<Result> = PromiseLike<Result> & {
+	cancel: (reason?: any) => void;
```
Comment (Samaritan1011001) [inline packages/storage/src/types/common.ts:23]: same here and other similar places

**Review by Samaritan1011001 — APPROVED** [review]

**Review by kvramyasri7 — APPROVED** [review]

File: packages/storage/src/providers/s3/apis/copy.ts (hunk 8 above)
Comment (AllanZhengYP) [reply packages/storage/src/providers/s3/apis/copy.ts:5]: That is true

**Pull request merged**